reporter/src/main/kotlin/utils/FreemarkerTemplateProcessor.kt

```diff
@@ -0,0 +1,242 @@
+            resolved.copy(licenses = resolved.licenses.sortedBy { it.license.toString() })
+        }
+
+        /**
+         * The resolved license file information for the package.
+         */
+        val licenseFiles: ResolvedLicenseFileInfo by lazy { input.licenseInfoResolver.resolveLicenseFiles(id) }
+
+        /**
+         * Return all [ResolvedLicense]s for this package excluding those licenses which are contained in any of the
+         * license files. This is useful when the raw texts of the license files are included in the generated output
+         * file and all licenses not contained in those files shall be listed separately.
+         */
+        @Suppress("UNUSED") // This function is used in the templates.
+        fun licensesNotInLicenseFiles(): List<ResolvedLicense> {
+            val outputFileLicenses = licenseFiles.files.flatMap { it.licenses }
+            return license.filter { it !in outputFileLicenses }
+        }
+    }
+
+    /**
+     * A collection of helper functions for the Freemarker templates.
+     */
+    class TemplateHelper(private val ortResult: OrtResult, private val licenseConfiguration: LicenseConfiguration) {
+        /**
+         * Return [packages] that are a dependency of at least one of the provided [projects][projectIds].
+         */
+        @Suppress("UNUSED") // This function is used in the templates.
+        fun filterByProjects(
+            packages: Collection<PackageModel>,
+            projectIds: Collection<Identifier>
+        ): List<PackageModel> {
+            val dependencies = projectIds.mapNotNull { ortResult.getProject(it) }
+                .flatMapTo(mutableSetOf()) { it.collectDependencies() }
+
+            return packages.filter { pkg -> pkg.id in dependencies }
+        }
+
+        /**
+         * Return licenses that are configured to be [included in a notice file][License.includeInNoticeFile] in the
+         * [LicenseConfiguration].
+         */
+        @Suppress("UNUSED") // This function is used in the templates.
+        fun filterIncludeInNoticeFile(licenses: Collection<ResolvedLicense>): List<ResolvedLicense> =
+            licenses.filter { resolvedLicense ->
+                licenseConfiguration.licenses.find { it.id == resolvedLicense.license }?.includeInNoticeFile ?: true
+            }
+
+        /**
+         * Return licenses that are configured to require a
+         * [source code offer in a notice file][License.includeSourceCodeOfferInNoticeFile] in the
+         * [LicenseConfiguration].
+         */
+        @Suppress("UNUSED") // This function is used in the templates.
+        fun filterIncludeSourceCodeOfferInNoticeFile(licenses: Collection<ResolvedLicense>): List<ResolvedLicense> =
+            licenses.filter { resolvedLicense ->
+                licenseConfiguration.licenses.find { it.id == resolvedLicense.license }
+                    ?.includeSourceCodeOfferInNoticeFile ?: false
+            }
```

Comment:
Note to myself: These helper functions still talk about NOTICE files. We probably should generalize the names, but this partly overlaps with https://github.com/oss-review-toolkit/ort/pull/3327. I believe we should first get this PR and @oheger-bosch's PR merged, and then do another small PR for a renaming.